### PR TITLE
Merge basic gtest unit test support into cmake files.

### PR DESCRIPTION
**Pull request by gfr10598**
Moves iproute2 cmake file into ext/iproute2, instead of in the main directory.
Introduces src/ directory.
Improves ordering of items in main CMakeLists.txt.

---

This change is [Reviewable](https://reviewable.io/reviews/m-lab/tcpinfo_lib/4)

### CODE REVIEW

File: ext/gtest/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 2.8.8)
```
Comment (stephen-soltesz) [inline ext/gtest/CMakeLists.txt]: Why do the minimum_required versions change across files? Is there a standard version we can adopt?

File: ext/gtest/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+project(gtest_builder C CXX)
+include(ExternalProject)
+
+#set(GTEST_FORCE_SHARED_CRT ON)
```
Comment (stephen-soltesz) [inline ext/gtest/CMakeLists.txt]: Are these comments necessary?

File: ext/iproute2/CMakeLists.txt [hunk 3]
```diff
@@ -24,6 +26,8 @@ ExternalProject_Add(
   TEST_COMMAND ""
 )
 
+ExternalProject_Get_Property(iproute2 source_dir)
```
Comment (stephen-soltesz) [inline ext/iproute2/CMakeLists.txt]: From other examples, it looks like the `ExternalProject_Get_Property` sets the variable `source_dir` but that variable isn't used in this file. If my understanding is incorrect, what is really happening? Or, is this directive not necessary for this file? Or, is there a missing use below?

File: ext/gtest/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#set(GTEST_DISABLE_PTHREADS OFF)
+
+ExternalProject_Add(googletest
```
Comment (stephen-soltesz) [inline ext/gtest/CMakeLists.txt:5]: It looks like `ext/gtest/CMakeLists.txt` and `ext/iproute2/CMakeLists.txt` use different conventions for placing the project name on the same line or following line. IMO, one style would be preferable to both. (The cmake docs kind of hint that this one might be the preferable style).

File: ext/iproute2/CMakeLists.txt [hunk 5]
```diff
@@ -9,11 +9,13 @@ ExternalProject_Add(
   UPDATE_COMMAND ""
   PATCH_COMMAND ""
 
-  SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/iproute2"
+  SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/build"
 
   CONFIGURE_COMMAND
       ./configure
 
+  # Unfortunately, we have to build in the source directory, because of
```
Comment (stephen-soltesz) [inline ext/iproute2/CMakeLists.txt]: Just clarifying: why is this unfortunate?

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,18 +1,49 @@
...
+  add_definitions(-Wall -ansi -Wno-deprecated -pthread)
+endif()
+
+
+################### gtest submodule ########################
+add_subdirectory(${PROJECT_SOURCE_DIR}/ext/gtest)
+enable_testing()
+set(PROJECT_TEST_NAME ${PROJECT_NAME_STR}_test)
+include_directories(${GTEST_INCLUDE_DIRS} ${COMMON_INCLUDES})
+
```
Comment (stephen-soltesz) [inline CMakeLists.txt:71]: Above this point, I think I gather what's going on: this is the top level cmake file that loads some libraries (probably not cmake's word for that) some predefined packages, as well as the custom packages(?) defined under the `ext/*` directories. Is this about right?

I like the section comment headers to delineate these blocks.

It also looks like a new block begins below, where all the above setup starts to be used, the googletest library in particular. What do you think of also adding a section header here to separate (what I'm imagining as) the "arrange" section above and the "action" section below?

**Review by stephen-soltesz — COMMENTED** [review]
The major caveat for this review is that I am not familiar with cmake.

I don't consider any of my comments blocking, but may help clarify some parts for a new reader of the code.

**Review by stephen-soltesz — APPROVED** [review]

**Comment by stephen-soltesz** [comment]
Discussed in person, and it sounds like more changes are on the way, so LGTM.

File: ext/gtest/CMakeLists.txt (hunk 1 above)
Comment (gfr10598) [reply ext/gtest/CMakeLists.txt]: I expect it is pretty random, but will settle on one common one.

File: ext/gtest/CMakeLists.txt (hunk 2 above)
Comment (gfr10598) [reply ext/gtest/CMakeLists.txt]: Nope.  Gone.

File: ext/gtest/CMakeLists.txt (hunk 4 above)
Comment (gfr10598) [reply ext/gtest/CMakeLists.txt:5]: Hmm - don't see anywhere where it is on the next line.  Trying to standardize on keeping the target(s) on the first line.

File: ext/iproute2/CMakeLists.txt (hunk 5 above)
Comment (gfr10598) [reply ext/iproute2/CMakeLists.txt]: Tweaked comment.

File: ext/iproute2/CMakeLists.txt (hunk 3 above)
Comment (gfr10598) [reply ext/iproute2/CMakeLists.txt]: Good catch.  Will probably need it later but it is not used (or exported) yet.

File: CMakeLists.txt (hunk 6 above)
Comment (gfr10598) [reply CMakeLists.txt:71]: Heh - already added these, but I guess I didn't push them yet.

**Pull request merged**